File: java/00895_maximum_frequency_stack.java
//       Java
// Time complexity - push, pop, peek: O (1)
// Space complexity - O (N)

// ["push", "push", "push", "push", "push", "push", "pop", "pop", "pop", "pop"]
// [ [5],    [7],    [5],    [7],    [4],    [5]],    5      7      5      4

// 5 - 3
// 7 - 2
// 4 - 1

class FreqStack {
    // Maintains frequencies of elements
    private Map<Integer, Integer> freqMap;
    // Maintains stacks of elements by frequency
    private Map<Integer, Stack<Integer>> groupMap;
    private int maxFreq = 0;

    public FreqStack() {
        freqMap = new HashMap<>();
        groupMap = new HashMap<>();
    }

    public void push(int x) {
        int freq = freqMap.getOrDefault(x, 0) + 1;
        freqMap.put(x, freq);

        maxFreq = Math.max(maxFreq, freq);

        if (!groupMap.containsKey(freq)) {
            groupMap.put(freq, new Stack<>());
        }
        groupMap.get(freq).push(x);
    }

    public int pop() {
        if (!groupMap.containsKey(maxFreq)) {
            return -1;
        }

        int val = groupMap.get(maxFreq).pop();
        freqMap.put(val, freqMap.get(val) - 1);

        if (groupMap.get(maxFreq).isEmpty()) {
            maxFreq--;
        }

        return val;
    }

    public int peek() {
        return groupMap.get(maxFreq).peek();
    }
}

File: python/00895_maximum_frequency_stack.py
# Time complexity - push, pop, peek: O (1)
# Space complexity - O (N)

# ["push", "push", "push", "push", "push", "push", "pop", "pop", "pop", "pop"]
# [ [5],    [7],    [5],    [7],    [4],    [5]],    5      7      5      4

# 5 - 3
# 7 - 2
# 4 - 1

# 3 - 5
# 2 - 7
# 1 - 4

class FreqStack:
    def __init__(self):
        self.freq_map = {}
        self.group_map = {}
        self.max_freq = 0

    def push(self, x: int) -> None:
        freq = self.freq_map.get(x, 0) + 1
        self.freq_map[x] = freq

        self.max_freq = max(self.max_freq, freq)

        if freq not in self.group_map:
            self.group_map[freq] = []

        self.group_map[freq].append(x)

    def pop(self) -> int:
        if self.max_freq in self.group_map and self.group_map[self.max_freq]:
            val = self.group_map[self.max_freq].pop()
            self.freq_map[val] -= 1

            if not self.group_map[self.max_freq]:
                self.max_freq -= 1

            return val
        return -1

    def peek(self) -> int:
        if self.max_freq in self.group_map and self.group_map[self.max_freq]:
            return self.group_map[self.max_freq][-1]
        return -1
